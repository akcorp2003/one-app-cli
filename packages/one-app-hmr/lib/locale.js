/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import path from 'path';
import chokidar from 'chokidar';

import {
  debug, log, warn, time, palegreen, orange, green, info,
} from './logs';
import {
  addLanguagePacksForModule,
  addModuleLanguagePack,
  removeModuleLanguagePack,
} from './utils';

export function printLocale(message) {
  return `${palegreen('locale')} - ${message}`;
}

export function printLocaleAction({ locale, moduleName, action }) {
  return log(printLocale(`${orange(`"${locale}"`)} for module ${orange(`"${moduleName}"`)} has been ${action}`));
}

export function getModuleInfoFromLocalePath(filePath) {
  const [modulePath, fileBasePath = ''] = filePath.split('/locale/');
  const [moduleName] = modulePath.split('/').reverse();
  const [languagePack] = fileBasePath.split('/');
  const locale = languagePack.toLowerCase().replace('.json', '');
  return [moduleName, modulePath, languagePack, locale];
}

export function loadLanguagePacksForModule(modulePath) {
  const [moduleName] = modulePath.split('/').reverse();
  const langPacksLoaded = addLanguagePacksForModule({ modulePath, moduleName });
  info(printLocale(`Loaded language packs for ${orange(`"${moduleName}"`)}: [ ${langPacksLoaded.map((langPack) => green(langPack)).join(', ')} ]`));
  return [moduleName, langPacksLoaded];
}

export function watchLanguagePackFileEvents(watcher, publish) {
  watcher
    .on('add', (filePath) => {
      const [moduleName, modulePath, locale] = getModuleInfoFromLocalePath(filePath);
      printLocaleAction({ locale, moduleName, action: 'added' });
      addModuleLanguagePack({
        filePath, moduleName, modulePath, locale,
      });
      publish({
        action: 'locale:add', path: filePath, moduleName, locale,
      });
    })
    .on('change', (filePath) => {
      const [moduleName, modulePath, locale] = getModuleInfoFromLocalePath(filePath);
      printLocaleAction({ locale, moduleName, action: 'changed' });
      addModuleLanguagePack({
        filePath, moduleName, modulePath, locale,
      });
      publish({
        action: 'locale:change', path: filePath, moduleName, locale,
      });
    })
    .on('unlink', (filePath) => {
      const [moduleName, modulePath, locale] = getModuleInfoFromLocalePath(filePath);
      printLocaleAction({ locale, moduleName, action: 'removed' });
      removeModuleLanguagePack({
        filePath, moduleName, modulePath, locale,
      });
      publish({
        action: 'locale:remove', path: filePath, moduleName, locale,
      });
    });
}

export function createHotLanguagePacks(modulePaths, publish) {
  const watcherOptions = { awaitWriteFinish: true };
  const watcherPaths = modulePaths.map((modulePath) => path.join(modulePath, 'locale'));
  const watcher = chokidar.watch(watcherPaths, watcherOptions);

  const moduleNames = modulePaths.map(loadLanguagePacksForModule).map(([moduleName]) => moduleName);

  watcher
    .on('error', (error) => warn(printLocale(`Language pack watcher error: ${error}`)))
    .on('ready', () => {
      log(printLocale(`Watching language packs for modules: [ ${moduleNames.map((name) => orange(JSON.stringify(name))).join(', ')} ]`));
      watchLanguagePackFileEvents(watcher, publish);
    });

  return watcher;
}

export async function loadLanguagePacks(app, { languagePacks, useLanguagePacks, publish }) {
  debug(printLocale('"useLanguagePacks" was set to "%s"'), useLanguagePacks);

  if (useLanguagePacks && languagePacks.length > 0) {
    debug(printLocale('Loading language packs %o'), languagePacks);
    await time(printLocale('Language pack setup time'), () => {
      createHotLanguagePacks(languagePacks, publish);
    });
  } else {
    debug(printLocale('Locale folders were not found'));
  }
}
