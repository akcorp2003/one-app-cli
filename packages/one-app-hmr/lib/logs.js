/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable no-console */

import createDebug from 'debug';
import chalk from 'chalk';

export const libName = 'one-app-hmr';

export const debug = createDebug(libName);

let logLevel = 2;
export function setLogLevel(level = 2) {
  logLevel = level;
}

export const {
  red, green, yellow,
} = chalk;
export const pink = chalk.keyword('pink');
export const bisque = chalk.keyword('bisque');
export const deeppink = chalk.keyword('deeppink');
export const dodgerblue = chalk.keyword('dodgerblue');
export const blueviolet = chalk.keyword('blueviolet');
export const palegreen = chalk.keyword('palegreen');
export const purple = chalk.keyword('purple');
export const orange = chalk.keyword('orange');
export const cyan = chalk.keyword('cyan');
export const printLibName = () => `  ${libName} ::`;
export const log = (message) => logLevel > 1 && console.log(`${cyan.bold(printLibName())} %s`, message);
export const warn = (message) => logLevel > 0 && console.warn(`${orange.bold(printLibName())} %s`, orange(message));
export const error = (message) => console.error(`${red.bold(printLibName())} %s`, red(message));
export const info = (message) => console.info(`\n${green.bold(printLibName())} %s`, green(message));
export const time = async (label, callback) => {
  const timeLabel = `${bisque.bold(printLibName())} ${label}`;
  console.time(timeLabel);
  await callback();
  console.timeEnd(timeLabel);
};

export function setDebugMode() {
  process.env.DEBUG = [process.env.DEBUG || '', 'one-app-hmr'].join(' ').trim();
}

setDebugMode();
