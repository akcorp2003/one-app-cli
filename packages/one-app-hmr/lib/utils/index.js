/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

export {
  isDevelopment,
} from './common';

export {
  vol,
  vfs,
} from './virtual-file-system';

export {
  getContextPath,
  getStaticPath,
  getModulesPath,
  getOneAppPath,
  getVendorsPath,
  getTempPath,
  combineUrlFragments,
  joinUrlFragments,
  getPublicUrl,
  getPublicModulesUrl,
  getPublicAppUrl,
  getPublicVendorsUrl,
} from './paths';

export {
  extractLanguageDataFromLocale,
  loadModuleLanguagePacks,
  addLanguagePacksForModule,
  addModuleLanguagePack,
  removeModuleLanguagePack,
} from './language-packs';

export {
  loadRemoteModuleMap,
  createLocalModuleMap,
  createUnifiedModuleMap,
  createModuleMap,
} from './module-map';

export {
  renderDocument,
  createInitialState,
  getLocalModulesFromStats,
  getRootModuleScriptUrl,
  getExternalsForScripts,
  getStatsFromResponse,
} from './rendering';

export {
  loadStatics,
  loadOneAppStaticsFromDocker,
  addStaticsDirToGitIgnore,
  preloadOneAppStatics,
} from './statics';
