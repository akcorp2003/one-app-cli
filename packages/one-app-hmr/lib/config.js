/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import fs from 'fs';
import path from 'path';
import readPkgUp from 'read-pkg-up';

import { getContext, getModulesPath, getPublicPath } from './webpack/utility';
import { debug } from './logs';

export const defaultConfig = {
  modules: [],
  externals: [],
  scenarios: [],
  locales: [],
  port: 4000,
  logLevel: 2,
  dockerImage: 'oneamex/one-app-dev',
  oneAppSource: 'docker', // "docker" or "git"
  rootModule: 'root-module',
  remoteModuleMap: null,
  useParrotMiddleware: true,
  useLanguagePacks: true,
};

export function getDefaultScenariosPath(modulePath = getContext()) {
  const scenariosPath = path.resolve(modulePath, 'mock', 'scenarios.js');
  if (fs.existsSync(scenariosPath)) return scenariosPath;
  return [];
}

export function getDefaultLocalesPath(modulePath = getContext()) {
  const localesPath = path.resolve(modulePath, 'locale');
  if (fs.existsSync(localesPath)) return modulePath;
  return [];
}

export const ONE_AMEX_CONFIG_KEY = 'one-amex';
export const BUNDLER_CONFIG_KEY = 'bundler';
export const RUNNER_CONFIG_KEY = 'runner';
export const HMR_CONFIG_KEY = 'hmr';

export async function getPackageJson(modulePath) {
  const {
    packageJson: {
      [ONE_AMEX_CONFIG_KEY]: oneAmexConfig,
      name: moduleName,
      version: moduleVersion,
    },
  } = await readPkgUp({ cwd: modulePath });
  return {
    oneAmexConfig,
    moduleName,
    moduleVersion,
  };
}

const getOneAmexConfig = (oneAmexConfig, moduleName) => {
  const {
    [RUNNER_CONFIG_KEY]: {
      modules, rootModuleName, moduleMapUrl, dockerImage,
    } = { modules: ['.'], rootModuleName: moduleName },
    [BUNDLER_CONFIG_KEY]: { providedExternals, requiredExternals } = {},
    [HMR_CONFIG_KEY]: hmr = {},
  } = oneAmexConfig || {};
  return {
    modules,
    rootModuleName,
    moduleMapUrl,
    dockerImage,
    hmr,
    providedExternals,
    requiredExternals,
  };
};
const getLocalModules = (modulePath, hmr, modules) => (hmr.modules || modules || [])
  .map((pathName) => path.resolve(modulePath, pathName))
  // .filter((pathName) => pathName !== modulePath)
  .filter((pathName) => fs.existsSync(pathName));
const getLanguagePacks = (modulePath, hmr) => []
  .concat(hmr.languagePacks || [], getDefaultLocalesPath(modulePath))
  .filter((pathName) => fs.existsSync(pathName));

const getScenarios = (modulePath, hmr) => []
  .concat(hmr.scenarios || [], getDefaultScenariosPath(modulePath)
  ).filter((pathName) => fs.existsSync(pathName));
const getExternals = (hmr, providedExternals, requiredExternals) => [...new Set([].concat(
  hmr.externals || [],
  providedExternals || [],
  requiredExternals || []
)).values()];
export async function getModuleConfig(modulePath = getContext()) {
  const { oneAmexConfig, moduleName, moduleVersion } = await getPackageJson(modulePath);
  const {
    modules, rootModuleName, moduleMapUrl, dockerImage, hmr, providedExternals, requiredExternals,
  } = getOneAmexConfig(oneAmexConfig, moduleName);

  const languagePacks = getLanguagePacks(modulePath, hmr);
  const scenarios = getScenarios(modulePath, hmr);
  const externals = getExternals(hmr, providedExternals, requiredExternals);
  return {
    moduleName,
    moduleVersion,
    modulePath,
    externals,
    scenarios,
    languagePacks,
    // hmr profile
    modules: getLocalModules(modulePath, hmr, modules),
    dockerImage: hmr.dockerImage || dockerImage,
    logLevel: hmr.logLevel || 2,
    port: hmr.port || 4000,
    useParrotMiddleware: hmr.useParrotMiddleware || true,
    useLanguagePacks: hmr.useLanguagePacks || true,
    oneAppSource: hmr.oneAppSource || 'docker', // "docker" or "git"
    rootModuleName: hmr.rootModuleName || rootModuleName,
    remoteModuleMap: hmr.remoteModuleMap || moduleMapUrl,
  };
}

export async function createConfig(initialConfig) {
  const baseConfig = initialConfig || await getModuleConfig();
  const {
    context = getContext(),
    publicPath = getPublicPath(),
    staticPath = getModulesPath(),
  } = baseConfig;
  const config = {
    ...defaultConfig,
    ...baseConfig,
    context,
    publicPath,
    staticPath,
  };

  const modules = await Promise.all(config.modules.map(getModuleConfig));
  modules.unshift(config);
  config.modules = modules;
  config.entryModule = config.moduleName;
  config.scenarios = []
    .concat(modules.map(({ scenarios }) => scenarios))
    .map((paths) => [].concat(paths).filter((pathName) => fs.existsSync(pathName)))
    .reduce((scenariosArray, nextSet) => scenariosArray.concat(nextSet), []);
  config.languagePacks = []
    .concat(modules.map(({ languagePacks }) => languagePacks))
    .map((paths) => [].concat(paths).filter((pathName) => fs.existsSync(pathName)))
    .reduce((langPacksArray, nextSet) => langPacksArray.concat(nextSet), []);
  config.externals = [
    ...new Set(
      modules.reduce((externalsArray, { externals }) => externalsArray.concat(externals), [])
    ).values(),
  ];

  debug('configuration %o', config.entryModule);

  return config;
}
