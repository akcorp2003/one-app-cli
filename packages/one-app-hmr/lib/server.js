/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import express from 'express';

import { createModuleMap, getPublicUrl, getStaticPath } from './utils';
import {
  loadWebpackMiddleware,
  createHotModuleRenderingMiddleware,
  createModulesProxyRelayMiddleware,
  loadParrotMiddleware,
} from './middleware';
import { loadLanguagePacks } from './locale';
import {
  info, warn, yellow, orange,
} from './logs';

export const acceptedMiddleware = (req, res) => {
  res.status(202);
};

export default async function hmrServer({
  port = 4000,
  context,
  publicPath,
  staticPath,
  rootModuleName,
  remoteModuleMap: remoteModuleMapUrl,
  modules,
  externals,
  scenarios,
  languagePacks,
  useParrotMiddleware,
  useLanguagePacks,
} = {}) {
  info('Starting HMR server');
  info(`Root Holocron module: ${orange(rootModuleName)}`);
  info(`Holocron modules loaded: ${modules.map(({ moduleName }) => orange(`"${moduleName}"`)).join(', ')}\n`);

  const {
    moduleMap,
    localModuleMap,
    remoteModuleMap,
  } = await createModuleMap({ modules, remoteModuleMapUrl });

  const serverAddress = `http://localhost:${port}/`;
  const {
    publish,
    devMiddleware,
    hotMiddleware,
  } = await loadWebpackMiddleware({
    context,
    publicPath,
    staticPath,
    modules,
    externals,
    rootModuleName,
    serverAddress,
  });

  const app = express();

  app
    .use(createModulesProxyRelayMiddleware({ moduleMap, localModuleMap, remoteModuleMap }))
    .use(devMiddleware)
    .use(hotMiddleware)
    .use(getPublicUrl(), express.static(getStaticPath()));

  await loadLanguagePacks(app, { languagePacks, useLanguagePacks, publish });
  await loadParrotMiddleware(app, { scenarios, useParrotMiddleware, publish });

  const renderMiddleware = await createHotModuleRenderingMiddleware({
    rootModuleName,
    moduleMap,
    errorReportingUrl: '/error',
  });

  app
    .get('*', renderMiddleware)
    .post('/error', acceptedMiddleware);

  return [app, app.listen(port, (error) => {
    if (error) throw error;
    info(`server is up on "${serverAddress}" - ${yellow('initializing HMR')}\n`);

    process.on('exit', () => {
      warn('server shutting down\n');
    });
  })];
}
