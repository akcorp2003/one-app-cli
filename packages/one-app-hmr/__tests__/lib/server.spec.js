/*
 * Copyright 2021 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { orange } from '../../lib/logs';
import sandboxServer from '../../lib/server';

jest.mock('../../lib/logs');
jest.mock('../../lib/utils/module-map', () => ({
  createModuleMap: async () => ({
    moduleMap: 'module-map.json',
    localModuleMap: '../sample-module/module-map.json',
    remoteModuleMap: 'https://one-app-statics.surge.sh/module-map.json',
  }),
}));
jest.mock('../../lib/middleware', () => ({
  loadWebpackMiddleware() {
    return {
      publish: jest.fn(),
      devMiddleware: {
        waitUntilValid: jest.fn(),
      },
      hotMiddleware: jest.fn(),
    };
  },
  createModulesProxyRelayMiddleware: jest.fn(),
  loadParrotMiddleware: jest.fn(),
  createHotModuleRenderingMiddleware: jest.fn(),
}));
jest.mock('../../lib/webpack');

jest.mock('express', () => {
  const mockedExpress = () => ({
    get: jest.fn(() => ({
      post: jest.fn(),
    })),
    use: jest.fn(() => ({
      use: jest.fn(() => ({
        use: jest.fn(() => ({
          use: jest.fn(),
        })),
      })),
    })),
    listen: jest.fn((listenArgs, callback) => {
      Promise.resolve().then(callback);
    }),
    static: jest.fn(),
  });
  Object.defineProperty(mockedExpress, 'static', { value: jest.fn() });
  return mockedExpress;
});

beforeAll(() => {
  orange.bold = jest.fn();
  jest.spyOn(console, 'info');
});

afterEach(() => {
  console.info.mockClear();
});

describe('sandboxServer', () => {
  const config = {
    port: 4000,
    context: jest.fn(),
    publicPath: '/publicPath/',
    staticPath: '/staticPath/',
    rootModuleName: 'root-module',
    remoteModuleMap: 'https://one-app-statics.surge.sh/module-map.json',
    modules: ['root-module', 'child-module'],
    externals: [],
    scenarios: [],
    languagePacks: [],
    useParrotMiddleware: true,
    useLanguagePacks: true,
  };
  it('hmr server starts up', async () => {
    const server = await sandboxServer(config);
    expect(server[0].listen).toHaveBeenCalled();
  });
});
