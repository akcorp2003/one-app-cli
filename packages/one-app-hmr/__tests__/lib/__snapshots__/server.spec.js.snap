// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hmrServer hmr server starts up 1`] = `Array []`;
